### PR TITLE
Powdered snow

### PR DESCRIPTION
**Pull request by CoderJoeW**
## Introduction
The goal of the PR is to introduce powder snow

https://github.com/pmmp/PocketMine-MP/projects/14#card-84464945

Adds the following

- New block "Powder Snow"
- New bucket "Powder Snow Bucket"
- Added cauldron integration with powder snow

### CODE REVIEW

**Review by dktapps — COMMENTED** [review]
You need to run `composer update-registry-annotations` to update the magic methods for VanillaBlocks and VanillaItems

File: src/block/PowderSnow.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+use pocketmine\item\Item;
+use pocketmine\item\VanillaItems;
+use pocketmine\world\sound\BucketEmptyPowderSnowSound;
+use pocketmine\world\sound\BucketFillPowderSnowSound;
+use pocketmine\world\sound\Sound;
+
+class PowderSnow extends Opaque{
+
+	public function getDropsForCompatibleTool(Item $item) : array{
```
Comment (diamond-gold) [inline src/block/PowderSnow.php]: Powder snow does not drop anything when mined

File: src/block/PowderSnow.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+		return [
+			VanillaItems::SNOWBALL()->setCount(4)
+		];
+	}
+
+	public function isAffectedBySilkTouch() : bool{
+		return true;
```
Comment (diamond-gold) [inline src/block/PowderSnow.php]: Powder snow does not drop anything when mined, even with Silk Touch

File: src/block/PowderSnowCauldron.php [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+use pocketmine\math\Vector3;
+use pocketmine\player\Player;
+use pocketmine\world\sound\CauldronEmptyPowderSnowSound;
+use pocketmine\world\sound\CauldronFillPowderSnowSound;
+use pocketmine\world\sound\Sound;
+
+final class PowderSnowCauldron extends FillableCauldron{
+
+	public function writeStateToWorld() : void{
+		parent::writeStateToWorld();
```
Comment (diamond-gold) [inline src/block/PowderSnowCauldron.php]: Unneeded override

File: src/block/PowderSnowCauldron.php [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+	public function getEmptySound() : Sound{
+		return new CauldronEmptyPowderSnowSound();
+	}
+
+	public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player $player = null, array &$returnedItems = []) : bool{
+		match($item->getTypeId()){
+			ItemTypeIds::BUCKET => $this->removeFillLevels(self::MAX_FILL_LEVEL, $item, VanillaItems::POWDER_SNOW_BUCKET(), $returnedItems),
+			ItemTypeIds::POWDER_SNOW_BUCKET, ItemTypeIds::WATER_BUCKET => $this->mix($item, VanillaItems::BUCKET(), $returnedItems),
+			ItemTypeIds::LINGERING_POTION, ItemTypeIds::POTION, ItemTypeIds::SPLASH_POTION => $this->mix($item, VanillaItems::GLASS_BOTTLE(), $returnedItems),
```
Comment (diamond-gold) [inline src/block/PowderSnowCauldron.php:53]: Unrelated change

File: src/world/World.php [hunk 5]
```diff
@@ -3003,7 +3003,7 @@ function() use ($resolver) : void{
 	 * @throws WorldException if the terrain is not generated
 	 */
 	public function getSafeSpawn(?Vector3 $spawn = null) : Position{
-		if(!($spawn instanceof Vector3) || $spawn->y < 1){
+		if(!($spawn instanceof Vector3) || $spawn->y <= $this->minY){
```
Comment (diamond-gold) [inline src/world/World.php:3006]: Unrelated change

File: src/block/PowderSnow.php [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+declare(strict_types=1);
+
+namespace pocketmine\block;
+
+use pocketmine\world\sound\BucketEmptyPowderSnowSound;
+use pocketmine\world\sound\BucketFillPowderSnowSound;
+use pocketmine\world\sound\Sound;
+
+class PowderSnow extends Opaque{
```
Comment (ShockedPlot7560) [inline src/block/PowderSnow.php:30]: Powder snow isn't an opaque bloc

Comment (ShockedPlot7560) [reply src/block/PowderSnow.php:30]: The block recovery logic via bucket is missing. The fact that the player slowly falls in with an effect is also missing.

File: src/item/BlockBucket.php [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+namespace pocketmine\item;
+
+use pocketmine\block\Block;
+use pocketmine\block\PowderSnow;
+use pocketmine\event\player\PlayerBucketEmptyEvent;
+use pocketmine\math\Vector3;
+use pocketmine\player\Player;
+
+class BlockBucket extends Item{
```
Comment (ShockedPlot7560) [inline src/item/BlockBucket.php:32]: By the same logic, lava and water should also be BlockBuckets. I don't think trying to generalize buckets is correct in this PR.

File: src/item/BlockBucket.php [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+	public function onInteractBlock(Player $player, Block $blockReplace, Block $blockClicked, int $face, Vector3 $clickVector, array &$returnedItems) : ItemUseResult{
+		if(!$blockReplace->canBeReplaced()){
+			return ItemUseResult::NONE();
+		}
+
+		//TODO: move this to generic placement logic
+		$resultBlock = clone $this->block;
+
+		if(!$resultBlock instanceof PowderSnow){
```
Comment (ShockedPlot7560) [inline src/item/BlockBucket.php:60]: Why name the BlockBucket class in this case and only accept snow, I think this class could just be renamed

**Review by ShockedPlot7560 — COMMENTED** [review]
This PR looks OK, but a lot of logic is missing to say it's implemented.

**Comment by ShockedPlot7560** [comment]
Ok I didn't see that it was an introduction, however I don't think it's the best thing to just put this half-implemented block.

File: resources/item_from_string_bc_map.json [hunk 9]
```diff
@@ -1328,6 +1328,7 @@
     "potato_block": "minecraft:potatoes",
     "potatoes": "minecraft:potatoes",
     "potion": "minecraft:potion",
+    "powder_snow": "minecraft:powder_snow",
```
Comment (dktapps) [inline resources/item_from_string_bc_map.json:1331]: This should not be changed - this file is kept only for BC purposes to support `LegacyStringToItemParser`

File: src/block/PowderSnow.php [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+	public function getBucketFillSound() : Sound{
```
Comment (dktapps) [inline src/block/PowderSnow.php:32]: Are these functions used by something? I don't remember them being implemented at the base.

Comment (ShockedPlot7560) [reply src/block/PowderSnow.php:32]: Its only used in `Blockbucket` class

Comment (dktapps) [reply src/block/PowderSnow.php:32]: Seems a bit redundant to have them here then ...
